design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,194 @@
+* Update existing built-in functions to use the TIMESTAMP data type
+* Casting between TIMESTAMP and BIGINT
+
+## What is not in scope
+* Changing the ROWTIME data type. We will eventually want this to happen, but that is a separate
+discussion.
+* Support for dates before Unix epoch - this is not supported by Kafka. There is a [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-228+Negative+record+timestamp+support)
+for negative timestamps, but until that is implemented, KSQL can only support positive timestamps.
+* DATE and TIME types - because TIMESTAMPs represent a point in time, DATE and TIME types
+would be useful if a user wants to represent a less specific time. These would be added after
+TIMESTAMP is implemented though.
+* Allow window units (HOUR, DAY etc) to be used in timestamp functions.
+
+## Public APIS
+
+The TIMESTAMP data type will store a point in time after Unix epoch without timezone information.
+The syntax is as follows:
+
+```roomsql
+CREATE STREAM stream_name (time TIMESTAMP, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field TIMESTAMP>) AS ...
+```
+
+TIMESTAMPS will be displayed in console as strings in ISO-8601 format:
+
+```roomsql
+> SELECT time FROM stream_name EMIT CHANGES;
+
++------------------------+
+|time                    |
++------------------------+
+|1994-11-05T13:15:30Z    |
+```
+
+TIMESTAMPS can be represented by milliseconds from Unix epoch or by using conversion functions:
+
+```roomsql
+INSERT INTO stream_name VALUES (1605927509166);
+INSERT INTO stream_name VALUES (STRINGTOTIMESTAMP("2020-11-20", "YYYY-MM-DD"));
+```
+
+## Design
+
+### Serialization/Deserialization
+
+TIMESTAMPs will be handled by the `java.time.Instant` class within KSQL. The corresponding Kafka Connect type is
+[org.apache.kafka.connect.data.Timestamp](https://kafka.apache.org/0100/javadoc/org/apache/kafka/connect/data/Timestamp.html).
+They are represented as long types in Schema Registry, but also come with a tag indicating that it
+is a timestamp, so they should be distinguishable from long types when handling serialized values in KSQL.
+
+#### Avro
+
+Avro schemas represent timestamps as
+```json
+{
+  "type": "long",
+  "logicalType": "timestamp-millis"
+}
+```
+The Avro deserializer KSQL uses supports this.
+
+#### Protobuf
+Protobuf 3 has a Timestamp type. The Protouf deserializer KSQL uses supports this.
+
+#### JSON/Delimited
+
+Timestamps will get stored in JSON and CSV files as long values. The KSQL JSON and delimited deserializers
+will be updated to parse timestamps.
+
+### UDFs
+
+The following UDFs should be updated to use the TIMESTAMP type instead of BIGINT:
+
+* TIMESTAMPTOSTRING
+* STRINGTOTIMESTAMP
+
+Because BIGINTs will be implicitly cast into TIMESTAMPs and vice versa, queries using these functions
+with BIGINT will still work when TIMESTAMP is introduced.
+
+There will also be a new function, `NOW()` that returns the current timestamp.
```

Comment:
Time is tricky in Kafka, may be beneficial to clarify which 'now' this would be. Is it the servers time or stream-time? If it is stream-time, will each partition have a different time? Will it be evaluated when the query is issued or when the function is called (important for persistent queries)?
All of those variants are useful in different scenarios and unfortunately all of them could be called `NOW`

Reply:
I agree with Andy's suggestion on this. I would also like to note that there is an existing function in KSQL called `UNIX_TIMESTAMP` that returns the time in ms when the function is called, so the distinction should be clearly defined in the docs.